Improve PHP 8.4 support

joaocosta-azores opened the pull request:
PHP 8.4 deprecates implicitly nullable types and applications are recommended to explicitly declare the type as nullable.
Message `Implicitly marking parameter $default as nullable is deprecated, the explicit nullable type must be used instead` is shown when declaring a default null value without declaring null in the type declaration.

To fix this, we use Union Type to declare the type also as nullable:
`string $value = null` becomes `string|null $value = null`

`composer run-script test` has been run to locate the code inside `src/` that needed to be fixed.

joaocosta-azores committed:
Fix nullable type declaration to support php 8.4

mjauvin commented:
If you could add php 8.4 in `.github/worflows/tests.yml` that would be awesome.

joaocosta-azores committed:
Add php 8.4 to tests.yml

bennothommo commented on src/Argon/Argon.php:
```diff
@@ -55,7 +55,7 @@ public static function parseWithCurrentLocale(
     public static function createFromFormatWithCurrentLocale(
         string $format,
-        string $time = null,
+        string|null $time = null,
```
I prefer this format - the only time I would use `|null` is at the end of a union type (ie. `string|int|null`).
```suggestion
        ?string $time = null,
```

bennothommo commented on src/Auth/AuthenticationException.php:
```diff
@@ -29,7 +29,7 @@ class AuthenticationException extends ApplicationException
      * @param Exception $previous Previous exception.
      */
-    public function __construct($message = "", $code = 0, Exception $previous = null)
+    public function __construct($message = "", $code = 0, Exception|null $previous = null)
```
```suggestion
    public function __construct($message = "", $code = 0, ?Exception $previous = null)
```

bennothommo commented on src/Exception/ApplicationException.php:
```diff
@@ -19,7 +19,7 @@ class ApplicationException extends ExceptionBase
      * @param Exception $previous Previous exception.
      */
-    public function __construct($message = "", $code = 0, Exception $previous = null)
+    public function __construct($message = "", $code = 0, Exception|null $previous = null)
```
```suggestion
    public function __construct($message = "", $code = 0, ?Exception $previous = null)
```

bennothommo commented on src/Exception/ExceptionBase.php:
```diff
@@ -51,7 +51,7 @@ class ExceptionBase extends Exception
      * @param Throwable $previous Previous exception.
      */
-    public function __construct($message = "", $code = 0, Throwable $previous = null)
+    public function __construct($message = "", $code = 0, Throwable|null $previous = null)
```
```suggestion
    public function __construct($message = "", $code = 0, ?Throwable $previous = null)
```

bennothommo commented on src/Exception/SystemException.php:
```diff
@@ -19,7 +19,7 @@ class SystemException extends ExceptionBase
      * @param Throwable $previous Previous exception.
      */
-    public function __construct($message = "", $code = 0, Throwable $previous = null)
+    public function __construct($message = "", $code = 0, Throwable|null $previous = null)
```
```suggestion
    public function __construct($message = "", $code = 0, ?Throwable $previous = null)
```

bennothommo commented on src/Halcyon/Datasource/Resolver.php:
```diff
@@ -31,7 +31,7 @@ public function __construct(array $datasources = [])
      * @inheritDoc
      */
-    public function datasource(string $name = null): DatasourceInterface
+    public function datasource(string|null $name = null): DatasourceInterface
```
```suggestion
    public function datasource(?string $name = null): DatasourceInterface
```

bennothommo commented on src/Halcyon/Datasource/ResolverInterface.php:
```diff
@@ -14,7 +14,7 @@ interface ResolverInterface
      * @throws \Winter\Storm\Halcyon\Exception\MissingDatasourceException If a datasource with the given name does not exist.
      */
-    public function datasource(string $name = null): DatasourceInterface;
+    public function datasource(string|null $name = null): DatasourceInterface;
```
```suggestion
    public function datasource(?string $name = null): DatasourceInterface;
```

bennothommo commented on src/Html/BlockBuilder.php:
```diff
@@ -103,7 +103,7 @@ public function append(string $name, string $content): void
      * If the block does not exist, then the `$default` content will be returned instead.
      */
-    public function placeholder(string $name, string $default = null): ?string
+    public function placeholder(string $name, string|null $default = null): ?string
```
```suggestion
    public function placeholder(string $name, ?string $default = null): ?string
```

bennothommo commented on src/Html/FormBuilder.php:
```diff
@@ -973,7 +973,7 @@ public function value($name, $value = null)
      * Returns a hidden HTML input, supplying the session key value.
      */
-    protected function requestHandler(string $name = null): string
+    protected function requestHandler(string|null $name = null): string
```
```suggestion
    protected function requestHandler(?string $name = null): string
```

bennothommo commented on src/Html/HtmlBuilder.php:
```diff
@@ -27,7 +27,7 @@ class HtmlBuilder
      * @return void
      */
-    public function __construct(UrlGenerator $url = null)
+    public function __construct(UrlGenerator|null $url = null)
```
```suggestion
    public function __construct(?UrlGenerator $url = null)
```

bennothommo commented on src/Mail/Mailer.php:
```diff
@@ -370,7 +370,7 @@ public function later($delay, $view, $data = null, $callback = null, $queue = nu
      * @return mixed
      */
-    public function laterOn($queue, $delay, $view, array $data = null, $callback = null)
+    public function laterOn($queue, $delay, $view, array|null $data = null, $callback = null)
```
```suggestion
    public function laterOn($queue, $delay, $view, ?array $data = null, $callback = null)
```

bennothommo commented on src/Support/helpers-array.php:
```diff
@@ -126,7 +126,7 @@ function array_except($array, $keys)
      * @return mixed
      */
-    function array_first($array, callable $callback = null, $default = null)
+    function array_first($array, callable|null $callback = null, $default = null)
```
```suggestion
    function array_first($array, ?callable $callback = null, $default = null)
```

bennothommo commented on src/Support/helpers-array.php:
```diff
@@ -141,7 +141,7 @@ function array_first($array, callable $callback = null, $default = null)
      * @return mixed
      */
-    function array_last($array, callable $callback = null, $default = null)
+    function array_last($array, callable|null $callback = null, $default = null)
```
```suggestion
    function array_last($array, ?callable $callback = null, $default = null)
```

joaocosta-azores committed:
Update src/Argon/Argon.php

Co-authored-by: Ben Thomson <git@alfreido.com>

joaocosta-azores committed:
Update src/Exception/ApplicationException.php

Co-authored-by: Ben Thomson <git@alfreido.com>

joaocosta-azores committed:
Update src/Auth/AuthenticationException.php

Co-authored-by: Ben Thomson <git@alfreido.com>

joaocosta-azores committed:
Update src/Exception/ExceptionBase.php

Co-authored-by: Ben Thomson <git@alfreido.com>

joaocosta-azores committed:
Update src/Exception/SystemException.php

Co-authored-by: Ben Thomson <git@alfreido.com>

joaocosta-azores committed:
Update src/Halcyon/Datasource/Resolver.php

Co-authored-by: Ben Thomson <git@alfreido.com>

joaocosta-azores committed:
Update src/Halcyon/Datasource/ResolverInterface.php

Co-authored-by: Ben Thomson <git@alfreido.com>

joaocosta-azores committed:
Update src/Html/BlockBuilder.php

Co-authored-by: Ben Thomson <git@alfreido.com>

joaocosta-azores committed:
Update src/Html/FormBuilder.php

Co-authored-by: Ben Thomson <git@alfreido.com>

joaocosta-azores committed:
Update src/Mail/Mailer.php

Co-authored-by: Ben Thomson <git@alfreido.com>

joaocosta-azores committed:
Update src/Support/helpers-array.php

Co-authored-by: Ben Thomson <git@alfreido.com>

joaocosta-azores committed:
Update src/Support/helpers-array.php

Co-authored-by: Ben Thomson <git@alfreido.com>

joaocosta-azores committed:
Update src/Html/HtmlBuilder.php

Co-authored-by: Ben Thomson <git@alfreido.com>

LukeTowers commented:
@bennothommo any idea why the one test is failing on 8.4? It seems very odd.

LukeTowers commented:
@bennothommo ping 😄

bennothommo committed:
Merge branch 'develop' into joaocosta-azores/develop

bennothommo commented:
Looks like PHP 8.4 has introduced a behaviour change, but I can't tell if it's in our code or Laravel's. Looking into it further.

bennothommo commented:
Turns out it was this:

![image](https://github.com/user-attachments/assets/1c9c381e-c98d-4921-981d-18bc611688b1)

We use a backtrace in `Winter\Storm\Database\Concerns\HasRelationships` to auto-guess the relation name in certain circumstances (such as changing a related ID directly). PHP 8.4 includes additional context for a closure in that backtrace which broke our auto-guessing method. I'll deploy a fix for this into @joaocosta-azores' PR.

bennothommo committed:
Update devcontainer to PHP 8.4

bennothommo committed:
Fix relation name auto-guessing in PHP 8.4

Per https://www.php.net/manual/en/migration84.new-features.php, debugging info for closures now include name, file and line of the closure, which means the caller function is no longer simply "{closure}" as it was pre-8.4, so we need to account for this signature change when scanning the backtrace to determine the relation name.

LukeTowers commented:
Holy that's a doozy of a hidden change. Thank goodness for tests 😂
